### PR TITLE
`matrix_rank_1_update`: kokkos impl

### PR DESCRIPTION
**Pull request by MikolajZuzek**
This is Kokkos implementation of `matrix_rank_1_update`, including conjugate overload for `matrix_rank_1_update_c`.
 #162  is a prerequisite for this, I can rebase once that is merged

### CODE REVIEW

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp [hunk 1]
```diff
@@ -0,0 +1,174 @@
+#ifndef LINALG_TPLIMPLEMENTATIONS_INCLUDE_EXPERIMENTAL___P1673_BITS_KOKKOSKERNELS_BLAS2_MATRIX_RANK_1_UPDATE_HPP_
```
Comment (mhoemmen) [inline tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp:1]: Would you consider adding a copyright header to the top of new files?

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+          for (idx_type i = 0; i < ext0; ++i) {
+            action(i, j);
+          }
+        });
+    }
+    exec.fence();
+  }
+
+private:
+  const ExecSpace &exec;
```
Comment (mhoemmen) [inline tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp:43]: Would you consider just holding all of these by value, not by reference?  Kokkos execution spaces are meant to be lightweight handles.  `MatrixType` is either an `mdspan` or `Kokkos::View`; both of those are view types, and thus idiomatically passed and held by value, not by reference.

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+         class Layout_A>
+void matrix_rank_1_update(kokkos_exec<ExecSpace> &&/* exec */,
+  std::experimental::mdspan<ElementType_x, std::experimental::extents<ext_x>, Layout_x,
+    std::experimental::default_accessor<ElementType_x>> x,
+  std::experimental::mdspan<ElementType_y, std::experimental::extents<ext_y>, Layout_y,
+    std::experimental::default_accessor<ElementType_y>> y,
+  std::experimental::mdspan<ElementType_A, std::experimental::extents<numRows_A, numCols_A>, Layout_A,
+    std::experimental::default_accessor<ElementType_A>> A)
+{
+  // constraints
```
Comment (mhoemmen) [inline tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp:95]: Constraints are enforced by, well, constraints, concepts, or SFINAE.  Mandates are enforced by `static_assert` or an equivalent.

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+  // preconditions
+  if ( A.extent(0) != x.extent(0) ){
+    throw std::runtime_error("KokkosBlas: matrix_rank_1_update: A.extent(0) != x.extent(0)");
+  }
+  if ( A.extent(1) != y.extent(0) ){
+    throw std::runtime_error("KokkosBlas: matrix_rank_1_update: A.extent(1) != y.extent(0)");
+  }
+
+#if defined KOKKOS_STDBLAS_ENABLE_TESTS
+  std::cout << "matrix_rank1_update: kokkos impl\n";
```
Comment (mhoemmen) [inline tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp:109]: Please see my comments on other PRs about logging encapsulation; thanks!

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp (hunk 2 above)
Comment (mhoemmen) [reply tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp:43]: More importantly, if you're capturing `MatrixType` by reference on host and trying to execute on device, you might have Problems : - ) .

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp (hunk 4 above)
Comment (MikolajZuzek) [reply tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp]: I've rebased this PR on on https://github.com/kokkos/stdBLAS/pull/170, which adds the function, and added https://github.com/kokkos/stdBLAS/pull/157/commits/377849734430a106b47ddfbe4b13b77f787aca79 to use it.

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp (hunk 1 above)
Comment (MikolajZuzek) [reply tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp:44]: Fixed on https://github.com/kokkos/stdBLAS/pull/157/commits/4142d9a1481873a53bfb7d5958016c5f1a89fa60

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp (hunk 2 above)
Comment (MikolajZuzek) [reply tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp]: Thank you, fixed on https://github.com/kokkos/stdBLAS/pull/157/commits/56c1c4cff96f93433a38ff1ab2fa3510c174aa45

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp (hunk 3 above)
Comment (MikolajZuzek) [reply tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp:148]: I added the mandates on https://github.com/kokkos/stdBLAS/pull/157/commits/fbc867da270bbcd83c7eb0c31da630e181afd64f, assuming we wish to explicitly check static extents match even if it's strictly implicated by dynamic extent match (checked in the preconditions) - let me know in other case.

I'll remove/replace `static_assert`s for the rank constraints once we conclude the discussion on https://github.com/kokkos/stdBLAS/pull/166#discussion_r846941134.

Comment (MikolajZuzek) [reply tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp]: Comment added on https://github.com/kokkos/stdBLAS/pull/157/commits/d935618e4af255e8c7f2e3c5af5b41402480f26c

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+namespace KokkosKernelsSTD {
+
+namespace Impl {
+
+// manages parallel execution of independent action
+// called like action(i, j) for each matrix element A(i, j)
+template <typename ExecSpace, typename MatrixType>
+class ParallelMatrixVisitor {
+public:
+  KOKKOS_INLINE_FUNCTION ParallelMatrixVisitor(ExecSpace &&exec_in, MatrixType &A_in):
```
[the remark this answers is not in the document]
Comment (MikolajZuzek) [reply tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp]: Thank you, fixed on https://github.com/kokkos/stdBLAS/pull/157/commits/dd2586959dedf5ac4fcef152dc1d56faf1d7bac9

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+          }
+        });
+    }
+    exec.fence();
+  }
+
+private:
+  const ExecSpace exec;
+  MatrixType A;
+  const size_t ext0;
```
[the remark this answers is not in the document]
Comment (MikolajZuzek) [reply tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp]: Fixed on https://github.com/kokkos/stdBLAS/pull/157/commits/dd2586959dedf5ac4fcef152dc1d56faf1d7bac9

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+          for (idx_type i = 0; i < ext0; ++i) {
+            action(i, j);
+          }
+        });
+    }
+    exec.fence();
+  }
+
+private:
+  const ExecSpace exec;
```
[the remark this answers is not in the document]
Comment (MikolajZuzek) [reply tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp]: Fixed on https://github.com/kokkos/stdBLAS/pull/157/commits/dd2586959dedf5ac4fcef152dc1d56faf1d7bac9

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+  MatrixType A;
+  const size_t ext0;
+  const size_t ext1;
```
[the remark this answers is not in the document]
Comment (MikolajZuzek) [reply tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp]: Fixed on https://github.com/kokkos/stdBLAS/pull/157/commits/dd2586959dedf5ac4fcef152dc1d56faf1d7bac9

**Review by mhoemmen — APPROVED** [review]
Excellent work!  Thanks for responding so quickly to feedback! : - D

**Pull request merged**